### PR TITLE
Changed WrapperState into an enum

### PR DESCRIPTION
**Pull request by amosonn**
`WrapperState` actually only holds one of its 3 fields at any time, therefore it is more naturally represented with an enum. As soon as it is changed from a `Callback` to an `Error` or a `Panic`, we also report an error to chmlib, so it should never reuse our callback.

### CODE REVIEW

File: chmlib/src/lib.rs [hunk 1]
```diff
@@ -172,32 +168,37 @@ where
...
+            // prevent double-frees).
+            let mut file = ManuallyDrop::new(ChmFile {
+                raw: NonNull::new_unchecked(file),
+            });
+            let unit = UnitInfo::from_raw(unit.read());
+            closure(&mut file, unit)
+        } else {
+            // if we once signaled failure, the callback shouldn't be reused
+            use std::hint::unreachable_unchecked;
+            unreachable_unchecked()
```
Comment (Michael-F-Bryan) [inline chmlib/src/lib.rs:185]: Isn't `unreachable_unchecked()` a little extreme?

**Review by Michael-F-Bryan — COMMENTED** [review]
Thanks for picking up that we're actually intending `WrapperState` to be an enum instead of a struct.

My only question is whether it's okay to use `unreachable_unchecked()`. If we're going to the effort of pattern matching on `WrapperState` We might as well handle the error case. Panicking with something like `panic!("You can't reuse the WrapperState")` would suffice, and then `catch_unwind()` will turn that into a `WrapperState::Panic`.

**Comment by amosonn** [comment]
Thanks for the swift reply! It's good that you left that part open in the blog-post, then I thought about it, solved with an enum, and went to compare to your solution :).

That `unreachable_unchecked` was obviously the one point I wasn't sure about either. First of all, it's worth mentioning that chmlib is definitely not expected to call the callback after failure. In the current version (with the struct) you might also run into trouble if that happens - if you panic once, "catch" that panic, and then chmlib calls the `FnMut` again, the internal state might be messed up; might not be UB, but the program may still act strange.
For me the deciding point for putting it there was looking at the source of chmlib and verifying that it indeed does the correct thing and exits once the callback fails.

However, if you decide that `unreachable_unchecked` is too extreme, panicing is probably still not the right solution; this will override the original error or panic, which stem from the actual user code and might be therefore more interesting for them, with a panic which originates from a bug in the library. The easiest drop-in solution would be to return `Continuation::Stop` which would silence the library error but report to the user their error properly. Maybe a better solution would be to add another variant, `Continuation::CalledAfterError` which would then lead to sending `chmlib_sys::CHM_ENUMERATOR_FAILURE` again, but without overriding the error. This can at some point also translate to some other variant of the `WrapperState`, e.g. `WrapperState::PanicAndCalledAgain(panic)`, which would also be propagated to the error returned from `handle_enumeration_result`. However, I think the deeper you go down that rabbit hole, the more of a mess it becomes - what's the right way to report to the user that they have an error/bug, but also the library they're using has one?

I actually thought this might be an interesting question to post on the forums. But to me it seems that `unreachable_unchecked` is eventually the correct solution - we can "prove" that this arm is never reached - when looking at the total source code, including that of chmlib.

**Comment by Michael-F-Bryan** [comment]
> For me the deciding point for putting it there was looking at the source of chmlib and verifying that it indeed does the correct thing and exits once the callback fails.

> However, I think the deeper you go down that rabbit hole, the more of a mess it becomes - what's the right way to report to the user that they have an error/bug, but also the library they're using has one?

> But to me it seems that unreachable_unchecked is eventually the correct solution - we can "prove" that this arm is never reached - when looking at the total source code, including that of chmlib.

After putting some thought into it, I think I agree with you here. While it feels a bit unsatisfying and potentially UB-inducing, by looking at the surrounding code and the way enumeration is done we've made sure it's impossible to ever reach this branch.

One alternative is to deliberately crash the program instead of the compiler potentially generating unsound code. After all, if the world is broken we should crash loudly so people can fix it. We could replace it with something like:

```rust
eprintln!("{}#{}: Enumeration was continued after an error or panic", file!(), line!());
eprintln!("This should be impossible and indicates the world is in a broken state, aborting...");
std::process::abort();
```

**Comment by amosonn** [comment]
Adding UB is the whole point of `unreachable_unchecked` ;) But more seriously, the main optimization I expect the compiler will be able to utilize here is removing the branch on the discriminant. This means that if this code runs after an error, it will most certainly fail spectacularly - if the layout of the different variants is different, then just segfaulting, but if it's the same, it will try to run data sitting behind the pointer as code.

This all boils down again to the question: How sure are we that that arm of the condition is never reached? If the answer is "completely", then `unreachable_unchecked` is the right choice. If not, then you come back to the question of what's the right way to report to the user that, in addition to an error-variant / panic-inducing-bug in their code, the library also did something surprising. Aborting the processes is just another possible answer to this question. I think in this case, the best solution is panicking, but while still preserving the original panic / error and passing them along.

Either way, I thought maybe other people might have something interesting to say about this, so I posted this also on the forums, here:
https://users.rust-lang.org/t/using-unreachable-unchecked-based-on-ffi-guarantees/35475
Let's say what comes up.

**Comment by Michael-F-Bryan** [comment]
After reading through the conversation on u.r-l.o I think I'm happy with using `unreachable_unchecked()` here. 

> In your situation you are feeding a callback and params to C, for it to call the function with the given params. Since the C function is able to call the callback ( *i.e.* , because it is an `unsafe fn` function pointer), then calling that C function must be considered `unsafe` too (which it will most probably be anyways since it is a foreign function).
>
> So when your other Rust code calls that C function, it needs to use an `unsafe` block to do it, and **it is there that your code must ensure the invariant is upheld** .

Since our `function_wrapper()` is already `unsafe` and we've made sure it's not possible to call the closure again after an error occurs, I think we should be okay.

Let's merge it :tada:

**Comment by amosonn** [comment]
Funny, I somehow got convinced of the opposite:

> You should treat `unreachable_unchecked` as an optimization.

And since this isn't performance-critical in any way, we should be Ok with just whatever happening there, perhaps even the panic you suggested at the first comment (since we're quite sure this isn't going to happen anyway).
But I guess it doesn't matter that much either way.

**Pull request merged**